tests/datasource/data_connector/test_configured_asset_azure_data_connector.py

```diff
@@ -0,0 +1,1346 @@
+container: my_container
+name_starts_with: my_base_directory/
+default_regex:
+   pattern: ^(.+)-(\\d{{4}})(\\d{{2}})\\.(csv|txt)$
+   group_names:
+       - data_asset_name
+       - year_dir
+       - month_dir
+assets:
+   alpha:
+       prefix: my_base_directory/alpha/files/go/here/
+       pattern: ^(.+)-(\\d{{4}})(\\d{{2}})\\.csv$
+   beta:
+       prefix: my_base_directory/beta_here/
+       pattern: ^(.+)-(\\d{{4}})(\\d{{2}})\\.txt$
+   gamma:
+       pattern: ^(.+)-(\\d{{4}})(\\d{{2}})\\.csv$
+
+   """
+    config = yaml.load(yaml_string)
+
+    mock_list_keys.return_value = [  # Initial return value during instantiation
+        "my_base_directory/alpha/files/go/here/alpha-202001.csv",
+        "my_base_directory/alpha/files/go/here/alpha-202002.csv",
+        "my_base_directory/alpha/files/go/here/alpha-202003.csv",
+        "my_base_directory/beta_here/beta-202001.txt",
+        "my_base_directory/beta_here/beta-202002.txt",
+        "my_base_directory/beta_here/beta-202003.txt",
+        "my_base_directory/beta_here/beta-202004.txt",
+        "my_base_directory/gamma-202001.csv",
+        "my_base_directory/gamma-202002.csv",
+        "my_base_directory/gamma-202003.csv",
+        "my_base_directory/gamma-202004.csv",
+        "my_base_directory/gamma-202005.csv",
+    ]
+
+    my_data_connector: ConfiguredAssetAzureDataConnector = (
+        instantiate_class_from_config(
+            config,
+            config_defaults={
+                "module_name": "great_expectations.datasource.data_connector"
+            },
+            runtime_environment={"name": "my_data_connector"},
+        )
+    )
+
+    # Since we are using mocks, we need to redefine the output of subsequent calls to `list_azure_keys()`
+    # Our patched object provides the ability to define a "side_effect", an iterable containing return
+    # values for subsequent calls. Since `_refresh_data_references_cache()` makes multiple calls to
+    # this method (once per asset), we define our expected behavior below.
+    #
+    # Source: https://stackoverflow.com/questions/24897145/python-mock-multiple-return-values
+
+    mock_list_keys.side_effect = [
+        [  # Asset alpha
+            "my_base_directory/alpha/files/go/here/alpha-202001.csv",
+            "my_base_directory/alpha/files/go/here/alpha-202002.csv",
+            "my_base_directory/alpha/files/go/here/alpha-202003.csv",
+        ],
+        [  # Asset beta
+            "my_base_directory/beta_here/beta-202001.txt",
+            "my_base_directory/beta_here/beta-202002.txt",
+            "my_base_directory/beta_here/beta-202003.txt",
+            "my_base_directory/beta_here/beta-202004.txt",
+        ],
+        [  # Asset gamma
+            "my_base_directory/gamma-202001.csv",
+            "my_base_directory/gamma-202002.csv",
+            "my_base_directory/gamma-202003.csv",
+            "my_base_directory/gamma-202004.csv",
+            "my_base_directory/gamma-202005.csv",
+        ],
+    ]
```

Comment:
This one test invokes `list_azure_keys` a few times. To ensure that we get different results with each call, we need to provide a `side_effect` list or iterable to tell the mock what to spit out at each step of the test.

The behavior here is very much aligned with S3 but I found it important to note this information for future users of this code.